### PR TITLE
Initial step bdf 2 3 refactor

### PR DESCRIPTION
**Pull request by Luckabarbeau**
Refactor the startup of bdf scheme. This allows sharp IB to use BDF 2 and BDF 3 scheme. There are now 3 options to start up these schemes. Multiple BDF step, which use a lower order BDF scheme to start, sdirk step that uses SDIRK to make the first few time steps, initial solution which use to initial solution to define the solution before the start of the simulation. 
This pull request also include 2 other small modifications:
Mini refactor of the linear solver try fail scheme.
A call to the base post-process_Fd for the sharp IB solver.

### CODE REVIEW

File: source/core/parameters.cc [hunk 1]
```diff
@@ -1030,6 +1055,11 @@ namespace Parameters
                         "1",
                         Patterns::Integer(),
                         "amg smoother overlap");
+      prm.declare_entry(
+        "block linear solver fail",
+        "false",
+        Patterns::Bool(),
+        "A bool that allow the use to block linear solver to throw errors");
```
Comment (blaisb) [inline source/core/parameters.cc:1062]: You should find a better name for this parameter. Like force linear solver continuation?
Something like that. Try to add a better description of it also :).

File: source/core/parameters.cc [hunk 2]
```diff
@@ -166,6 +185,11 @@ namespace Parameters
                         Patterns::Selection("none|iteration|end"),
                         "Clock monitoring methods "
                         "Choices are <none|iteration|end>.");
+      prm.declare_entry(
+        "write time in error table",
+        "false",
+        Patterns::Bool(),
+        "Bool to define if the time is written in the error table");
```
Comment (oguevremont) [inline source/core/parameters.cc:192]:
```suggestion
        "Boolean to define if the time is written in the error table");
```

File: source/core/simulation_control.cc [hunk 3]
```diff
@@ -48,6 +55,116 @@ SimulationControl::is_output_iteration()
...
+void
+SimulationControl::update_assembly_method()
+{
+  if (iteration_number <= 1 &&
+      method == Parameters::SimulationControl::TimeSteppingMethod::bdf2 &&
+      bdf_start_method ==
+        Parameters::SimulationControl::BDFStartupMethods::multiple_step_bdf)
+    {
+      assembly_method = Parameters::SimulationControl::TimeSteppingMethod::bdf1;
+      set_current_time_step(initial_time_step * startup_timestep_scaling);
```
Comment (blaisb) [inline source/core/simulation_control.cc:67]: One thing we could consider would be to abandon this startup_timestep_scaling parameter in a next PR and just use the regular time step and live with it.

File: source/core/parameters.cc (hunk 1 above)
Comment (oguevremont) [inline source/core/parameters.cc:1062]:
```suggestion
        "A boolean that allows the used block linear solver to throw errors");
```

File: source/core/simulation_control.cc [hunk 4]
```diff
@@ -48,6 +55,116 @@ SimulationControl::is_output_iteration()
...
+           bdf_start_method ==
+             Parameters::SimulationControl::BDFStartupMethods::sdirk_step)
+    {
+      return Parameters::SimulationControl::TimeSteppingMethod::sdirk33;
+    }
+  else
+    {
+      return method;
+    }
+}
```
Comment (blaisb) [inline source/core/simulation_control.cc:166]: I don't understand, why did you need an update_method and a get_method that both manage a bit the same thing? is it because of the time-step scaling?

File: source/core/simulation_control.cc [hunk 5]
```diff
@@ -48,6 +55,116 @@ SimulationControl::is_output_iteration()
...
+             Parameters::SimulationControl::BDFStartupMethods::sdirk_step)
+    {
+      assembly_method =
+        Parameters::SimulationControl::TimeSteppingMethod::sdirk22;
+      set_suggested_time_step(initial_time_step);
+    }
+  else
+    {
+      assembly_method = method;
+    }
```
Comment (oguevremont) [inline source/core/simulation_control.cc:142]: I think cascading "if"s would improve the readability, for example with level 1 - BDFStartupMethods, level 2 - TimeSteppingMethod, level 3 - iteration number.

File: source/solvers/gls_sharp_navier_stokes.cc [hunk 6]
```diff
@@ -634,6 +635,12 @@ GLSSharpNavierStokesSolver<dim>::postprocess_fd(bool firstIter)
       this->write_output_results(this->present_solution);
     }
 
+  bool enable =
+    this->simulation_parameters.analytical_solution->calculate_error();
+  this->simulation_parameters.analytical_solution->set_enable(false);
+  NavierStokesBase<dim, TrilinosWrappers::MPI::Vector, IndexSet>::
+    postprocess_fd(firstIter);
```
Comment (blaisb) [inline source/solvers/gls_sharp_navier_stokes.cc:642]: What is this? You do this to use the postprocess_fd without having it calculate error? 
Could you just not overload the calculate_L2_error and that would work? or does it not because the function is called from the base class?

File: source/solvers/navier_stokes_base.cc [hunk 7]
```diff
@@ -535,117 +532,11 @@ NavierStokesBase<dim, VectorType, DofsType>::iterate()
...
-      this->simulation_control->set_assembly_method(
-        Parameters::SimulationControl::TimeSteppingMethod::bdf3);
-      PhysicsSolver<VectorType>::solve_non_linear_system(false);
-
-      multiphysics->solve(
-        Parameters::SimulationControl::TimeSteppingMethod::bdf3);
-
-      simulation_control->set_suggested_time_step(timeParameters.dt);
-    }
-}
```
Comment (blaisb) [inline source/solvers/navier_stokes_base.cc:648]: So happy to see this gone. Was crap code and my entire fault.

File: tests/core/simulation_control_05.cc [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+  {
+    simulation_control_parameters.method =
+      Parameters::SimulationControl::TimeSteppingMethod::bdf2;
+    SimulationControlTransient simulation_control(
+      simulation_control_parameters);
+
+
+    // Constant time-stepping
+    deallog << "*************************************************" << std::endl;
+    deallog << "Constant time stepping - constant output" << std::endl;
```
Comment (blaisb) [inline tests/core/simulation_control_05.cc:59]: You should change this message to indicate what it is that you are testing... no?

File: tests/core/simulation_control_05.cc [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+ * ---------------------------------------------------------------------
+
+*
+* Author: Bruno Blais, Polytechnique Montreal, 2020-
+*/
+
+/**
+ * @brief This test checks that the time-step round-off errors
+ * do not affect the end of the simulation
+ */
```
Comment (blaisb) [inline tests/core/simulation_control_05.cc:23]: DId you modify an existing or create a new one? Change the header and add an explanation of what the test does.

File: tests/core/simulation_control_05.output [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+DEAL::Iteration : 3    Time : 3.00000
+DEAL::Iteration : 4    Time : 4.00000
+DEAL::*************************************************
+DEAL::Constant time stepping - constant output
+DEAL::*************************************************
+DEAL::Iteration : 0    Time : 0.00000
+DEAL::Iteration : 1    Time : 1.00000
+DEAL::Iteration : 2    Time : 2.00000
+DEAL::Iteration : 3    Time : 3.00000
+DEAL::Iteration : 4    Time : 4.00000
```
Comment (blaisb) [inline tests/core/simulation_control_05.output:36]: This output will not help you debug the test if it crashes. You should have it print something worthwhile (like e.g the method used for the assembly)

**Review by blaisb — COMMENTED** [review]
Good work. There are still some things left to address and fix (e.g. some tests and etc.). Your unit test is not good right now, it needs to test what you want to test.
You forgot some comments there and there also.

File: source/solvers/gls_sharp_navier_stokes.cc [hunk 11]
```diff
@@ -667,7 +674,21 @@ GLSSharpNavierStokesSolver<dim>::postprocess_fd(bool firstIter)
...
+              double total_time = 0;
+              for (auto it = summary.begin(); it != summary.end(); ++it)
+                {
+                  total_time += summary[it->first];
+                }
+              this->error_table.add_value("total_time", total_time);
+            }
         }
+      // Calculate pressure drop between two boundaries
+
```
Comment (oguevremont) [inline source/solvers/gls_sharp_navier_stokes.cc:691]:
```suggestion
```

**Review by oguevremont — APPROVED** [review]
Nice work! A few comments missing here and there, but overall clear. The cascading if statements is a suggestion I got on another PR and I think it would apply well in this case, but it is not absolutely necessary.

File: include/core/sdirk.h [hunk 12]
```diff
@@ -76,12 +76,14 @@ number_of_intermediary_stages(
   if (method == Parameters::SimulationControl::TimeSteppingMethod::sdirk33_1 ||
       method == Parameters::SimulationControl::TimeSteppingMethod::sdirk33_2 ||
       method == Parameters::SimulationControl::TimeSteppingMethod::sdirk33_3 ||
-      method == Parameters::SimulationControl::TimeSteppingMethod::sdirk33)
+      method == Parameters::SimulationControl::TimeSteppingMethod::sdirk33 ||
+      method == Parameters::SimulationControl::TimeSteppingMethod::bdf3)
```
[the remark this answers is not in the document]
Comment (Luckabarbeau) [reply include/core/sdirk.h:80]: BDF2 and BDF3 now require stage if they start up with sdirk so should i move that to another function i use that to initialize the solution stage vector for this case.

File: source/core/simulation_control.cc (hunk 4 above)
Comment (Luckabarbeau) [reply source/core/simulation_control.cc:166]: the two have been merge together

File: source/solvers/gls_sharp_navier_stokes.cc (hunk 6 above)
Comment (Luckabarbeau) [reply source/solvers/gls_sharp_navier_stokes.cc:642]: It is not possible to have calculate_L2_error virtual since it is from a base class

File: source/core/simulation_control.cc (hunk 5 above)
Comment (Luckabarbeau) [reply source/core/simulation_control.cc]: This would lead to duplication of the else statement so for the moment i think it would let it as it is for the time being.

File: source/core/parameters.cc [hunk 13]
```diff
@@ -123,6 +131,17 @@ namespace Parameters
         {
           std::runtime_error("Invalid time stepping scheme");
         }
+      const std::string bdf_startup_string = prm.get("bdf startup method");
+      if (bdf_startup_string == "multiple step bdf")
+        bdf_startup_method = BDFStartupMethods::multiple_step_bdf;
+      else if (bdf_startup_string == "sdirk step")
```
[the remark this answers is not in the document]
Comment (Luckabarbeau) [reply source/core/parameters.cc:137]: i would avoid using the bdf alone as it is used in other places in the code with a different signification (same fo sdirk).

[the remark this answers is not in the document]
Comment (blaisb) [reply source/core/parameters.cc:137]: Good argument, agreed

**Review by blaisb — APPROVED** [review]
